IOBalancerMemoryLeakTest.testMemoryLeak_with_SocketConnections

```
java.lang.AssertionError: expected:<0> but was:<20>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at com.hazelcast.nio.tcp.nonblocking.iobalancer.IOBalancerMemoryLeakTest$3.run(IOBalancerMemoryLeakTest.java:125)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:851)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:865)
	at com.hazelcast.nio.tcp.nonblocking.iobalancer.IOBalancerMemoryLeakTest.testMemoryLeak_with_SocketConnections(IOBalancerMemoryLeakTest.java:114)
```
https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.maintenance-nightly/com.hazelcast$hazelcast/780/testReport/junit/com.hazelcast.nio.tcp.nonblocking.iobalancer/IOBalancerMemoryLeakTest/testMemoryLeak_with_SocketConnections/

https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.x-nightly/com.hazelcast$hazelcast/880/testReport/junit/com.hazelcast.nio.tcp.nonblocking.iobalancer/IOBalancerMemoryLeakTest/testMemoryLeak_with_SocketConnections/

--- a/hazelcast/src/main/java/com/hazelcast/nio/tcp/nonblocking/iobalancer/IOBalancer.java
+++ b/hazelcast/src/main/java/com/hazelcast/nio/tcp/nonblocking/iobalancer/IOBalancer.java
@@ -18,6 +18,8 @@ package com.hazelcast.nio.tcp.nonblocking.iobalancer;
 
 import com.hazelcast.instance.HazelcastThreadGroup;
 import com.hazelcast.internal.metrics.Probe;
+import com.hazelcast.internal.util.counters.MwCounter;
+import com.hazelcast.internal.util.counters.SwCounter;
 import com.hazelcast.logging.ILogger;
 import com.hazelcast.logging.LoggingService;
 import com.hazelcast.nio.tcp.TcpIpConnection;
@@ -25,8 +27,6 @@ import com.hazelcast.nio.tcp.nonblocking.MigratableHandler;
 import com.hazelcast.nio.tcp.nonblocking.NonBlockingIOThread;
 import com.hazelcast.nio.tcp.nonblocking.NonBlockingSocketReader;
 import com.hazelcast.nio.tcp.nonblocking.NonBlockingSocketWriter;
-import com.hazelcast.internal.util.counters.MwCounter;
-import com.hazelcast.internal.util.counters.SwCounter;
 
 import static com.hazelcast.instance.GroupProperty.IO_BALANCER_INTERVAL_SECONDS;
 import static com.hazelcast.instance.GroupProperty.IO_THREAD_COUNT;
@@ -103,13 +103,17 @@ public class IOBalancer {
     }
 
     public void connectionAdded(TcpIpConnection connection) {
-        inLoadTracker.notifyConnectionAdded(connection);
-        outLoadTracker.notifyConnectionAdded(connection);
+        NonBlockingSocketReader socketReader = (NonBlockingSocketReader) connection.getSocketReader();
+        NonBlockingSocketWriter socketWriter = (NonBlockingSocketWriter) connection.getSocketWriter();
+        inLoadTracker.notifyHandlerAdded(socketReader);
+        outLoadTracker.notifyHandlerAdded(socketWriter);
     }
 
     public void connectionRemoved(TcpIpConnection connection) {
-        inLoadTracker.notifyConnectionRemoved(connection);
-        outLoadTracker.notifyConnectionRemoved(connection);
+        NonBlockingSocketReader socketReader = (NonBlockingSocketReader) connection.getSocketReader();
+        NonBlockingSocketWriter socketWriter = (NonBlockingSocketWriter) connection.getSocketWriter();
+        inLoadTracker.notifyHandlerRemoved(socketReader);
+        outLoadTracker.notifyHandlerRemoved(socketWriter);
     }
 
     public void start() {

--- a/hazelcast/src/main/java/com/hazelcast/nio/tcp/nonblocking/iobalancer/LoadTracker.java
+++ b/hazelcast/src/main/java/com/hazelcast/nio/tcp/nonblocking/iobalancer/LoadTracker.java
@@ -17,11 +17,8 @@
 package com.hazelcast.nio.tcp.nonblocking.iobalancer;
 
 import com.hazelcast.logging.ILogger;
-import com.hazelcast.nio.tcp.TcpIpConnection;
 import com.hazelcast.nio.tcp.nonblocking.MigratableHandler;
 import com.hazelcast.nio.tcp.nonblocking.NonBlockingIOThread;
-import com.hazelcast.nio.tcp.nonblocking.NonBlockingSocketReader;
-import com.hazelcast.nio.tcp.nonblocking.NonBlockingSocketWriter;
 import com.hazelcast.util.ItemCounter;
 import java.util.HashMap;
 import java.util.HashSet;
@@ -35,7 +32,7 @@ import static com.hazelcast.util.StringUtil.LINE_SEPARATOR;
 
 /**
  * Tracks the load of of NonBlockingIOThread(s) and creates a mapping between NonBlockingIOThread -> Handler.
- *
+ * <p/>
  * This class is not thread-safe with the exception of
  * {@link #addHandler(MigratableHandler)}   and
  * {@link #removeHandler(MigratableHandler)}
@@ -138,14 +135,14 @@ class LoadTracker {
         }
     }
 
-    public void notifyConnectionAdded(TcpIpConnection connection) {
-        ConnectionAddedTask connectionAddedTask = new ConnectionAddedTask(connection);
-        tasks.offer(connectionAddedTask);
+    public void notifyHandlerAdded(MigratableHandler handler) {
+        AddHandlerTask addHandlerTask = new AddHandlerTask(handler);
+        tasks.offer(addHandlerTask);
     }
 
-    public void notifyConnectionRemoved(TcpIpConnection connection) {
-        ConnectionRemovedTask connectionRemovedTask = new ConnectionRemovedTask(connection);
-        tasks.offer(connectionRemovedTask);
+    public void notifyHandlerRemoved(MigratableHandler handler) {
+        RemoveHandlerTask removeHandlerTask = new RemoveHandlerTask(handler);
+        tasks.offer(removeHandlerTask);
     }
 
 
@@ -257,47 +254,41 @@ class LoadTracker {
         sb.append(LINE_SEPARATOR);
     }
 
-    class ConnectionRemovedTask implements Runnable {
+    class RemoveHandlerTask implements Runnable {
 
-        private TcpIpConnection connection;
+        private final MigratableHandler handler;
 
-        public ConnectionRemovedTask(TcpIpConnection connection) {
-            this.connection = connection;
+        public RemoveHandlerTask(MigratableHandler handler) {
+            this.handler = handler;
         }
 
         @Override
         public void run() {
-            NonBlockingSocketReader socketReader = (NonBlockingSocketReader) connection.getSocketReader();
-            NonBlockingSocketWriter socketWriter = (NonBlockingSocketWriter) connection.getSocketWriter();
 
             if (logger.isFinestEnabled()) {
-                logger.finest("Removing handlers from: " + connection);
+                logger.finest("Removing handler : " + handler);
             }
 
-            removeHandler(socketReader);
-            removeHandler(socketWriter);
+            removeHandler(handler);
         }
     }
 
-    class ConnectionAddedTask implements Runnable {
+    class AddHandlerTask implements Runnable {
 
-        private TcpIpConnection connection;
+        private final MigratableHandler handler;
 
-        public ConnectionAddedTask(TcpIpConnection connection) {
-            this.connection = connection;
+        public AddHandlerTask(MigratableHandler handler) {
+            this.handler = handler;
         }
 
         @Override
         public void run() {
-            NonBlockingSocketReader socketReader = (NonBlockingSocketReader) connection.getSocketReader();
-            NonBlockingSocketWriter socketWriter = (NonBlockingSocketWriter) connection.getSocketWriter();
 
             if (logger.isFinestEnabled()) {
-                logger.finest("Added handlers for: " + connection);
+                logger.finest("Adding handler : " + handler);
             }
 
-            addHandler(socketReader);
-            addHandler(socketWriter);
+            addHandler(handler);
         }
     }
 

